### DISCUSSION TITLE
Optuna doesn't consume 100% of my CPUs

### DISCUSSION THREAD

**Discussion by ibobak** [optuna/optuna#4884] [Q&A]
### Expected behavior

While training CatBoost model using Optuna, I am trying to make so that all my CPUs are used.  But unfortunately (see the screenshot below) not even half of CPUs are used.

My question is:  why Optuna cannot load all CPUs? What is the problem? 

Code:
```
# there are 32 columns in my training dataframe and about 1000 rows. 
# if CatBoost model is trained separately (without Optuna), the training process works for about 10 seconds. 

ds_tr = Pool(data=mat_train, label=y_train, cat_features=cols_features_cat)
ds_te = Pool(data=mat_test, label=y_test, cat_features=cols_features_cat)

def calc_objective(trial):
    params = {
        "random_seed": 42,
        "thread_count": -1,
        "task_type": "CPU",
        #"custom_metric": ["MAPE"],
        "use_best_model": True,

        "iterations": trial.suggest_categorical("iterations", [300, 600, 1000]),
        
        "boosting_type": trial.suggest_categorical("boosting_type", ["Ordered", "Plain"]),
        "bootstrap_type": trial.suggest_categorical(
            "bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]
        ),
        "rsm": trial.suggest_float("colsample_bylevel", 0.1, 1),  # =colsample_bylevel
        "depth":  trial.suggest_int("depth", 2, 6),
        "learning_rate": trial.suggest_float("learning_rate", 0.005, 0.2),
        "l2_leaf_reg": trial.suggest_int("l2_leaf_reg", 1, 5),
        "random_strength": trial.suggest_float("random_strength", 0.2, 4)
    }
    
    if params["bootstrap_type"] == "Bayesian":
        params["bagging_temperature"] = trial.suggest_float("bagging_temperature", 0, 10)
        
    elif params["bootstrap_type"] == "Bernoulli":
        params["subsample"] = trial.suggest_float("subsample", 0.1, 1)
    
    model = CatBoostRegressor(**params)
    model.fit(
        ds_tr,
        eval_set=ds_te,
        early_stopping_rounds=100,
        plot=False,
        verbose=False
    )
    return model.best_score_["validation"]["RMSE"]


storage = optuna.storages.RDBStorage(
    url='postgresql+psycopg2://postgres:123@localhost:5432/optuna',
    engine_kwargs={'pool_size': 44, 'max_overflow': 0}
)

name_of_study = 'catboost13'
study = optuna.create_study(study_name=name_of_study, storage=storage, direction='minimize', load_if_exists=False)

study.optimize(calc_objective, n_trials=60000, n_jobs=44, timeout=3600*6)
```

What else I tried to modify in my code:
* I tried to put "thread_count": 1    for Catboost - to avoid Catboost using multiple cores and to allow Optuna to parallelize those things which are one-threaded.   That did not help - it was even worse picture with CPU utilization
* I tried to replace  the "study.optimize" call with this:
   ```
   from joblib import parallel_backend
   with parallel_backend('multiprocessing'):  # Overrides `prefer="threads"` to use multi-processing.
       study.optimize(calc_objective, n_trials=2000, n_jobs=44, timeout=600)
   ```
as it was suggested here https://github.com/optuna/optuna/issues/2202 , but that did not help either.

![image](https://github.com/optuna/optuna/assets/8807043/b69e2b3e-290f-4539-866c-4367dbe51c14)

Note: I also thought that postgresql and my harddrive may be the factor of slowdown, but this is not so - the diskgraph shows it is used on the speed of 1-2 MB/sec which is far below its possibility, and besides, I have a plenty of hardware to process these requests. 

### Environment

- Hardware:   2xCPU  Xeon 2696v4, 256GB RAM, 2TB M2 SSD.  
- Optuna version:3.3.0
- Python version:3.10.12
- OS: Ubuntu 22.04
- (Optional) Other libraries and their versions:
my requirements.txt (with which I've built the environment prior to installing optuna) was this:

numpy==1.23.5
pandas==1.5.3
pandas_datareader==0.10.0
SQLAlchemy==1.4.49
pandasql==0.7.3
seaborn==0.12.2
redshift_connector==2.0.911
apache-airflow==2.5.3
psycopg2-binary==2.9.7
tsfresh==0.20.1
xgboost==1.7.6
lightgbm==4.0.0
prophet==1.1.4
catboost==1.2
pymc==5.6.0
arviz==0.15.1
shap==0.42.1
ipython==8.10.0
ipykernel==6.19.2
ipywidgets==7.6.5
ipynb-py-convert==0.4.6
notebook==6.5.2
jupyter_contrib_nbextensions==0.7.0
jupyter_client==7.4.4
tornado==6.2
keyring==23.4.0
mlflow==2.5.0
coloredlogs==15.0.1
tenacity==8.0.1
lazyprofiler==0.1.1
boto3==1.28.29
smart_open==6.3.0

### Error messages, stack traces, or logs

```shell
-
```

### Steps to reproduce

(source code provided above)

**Comment by turbotimon** [comment]
Unfortunately optuna `n_jobs` uses multithreading instead of multiprocessing...

See https://github.com/optuna/optuna/issues/497

How to do multithreading: https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html
